src/nupic/algorithms/Cells4.cpp

```diff
@@ -1628,46 +1667,55 @@ void Cells4::adaptSegment(const SegmentUpdate& update)
     // updateSynapses? It would be cleaner and avoid the need to keep the
     // synapses sorted.
 
-    // Accumulate list of synapses to decrement, increment, add, and remove
-    std::set<UInt> synapsesSet(update.begin(), update.end());
+    // Accumulate lists of synapses to decrement, increment, add, and remove
+    std::set<UInt> synapsesSet(update.begin(), update.end()); // srcCellIdx
     static std::vector<UInt> removed; // srcCellIdx
-    static std::vector<UInt> synToDec, synToInc,
-                             inactiveSegmentIndices, activeSegmentIndices;
-    removed.clear() ;                       // purge residual data
-    synToDec.clear() ;                      // purge residual data
-    synToInc.clear() ;                      // purge residual data
-    inactiveSegmentIndices.clear() ;        // purge residual data
-    activeSegmentIndices.clear() ;          // purge residual data
-    for (UInt i = 0; i != segment.size(); ++i) {
-      UInt srcCellIdx = segment[i].srcCellIdx();
-      if (not_in(srcCellIdx, synapsesSet)) {
-        synToDec.push_back(srcCellIdx);  // TODO: Check synapse still exists!
-        inactiveSegmentIndices.push_back(i);
-      }
-      else {
-        synToInc.push_back(srcCellIdx);
-        synapsesSet.erase(srcCellIdx);
-        activeSegmentIndices.push_back(i);
-      }
-    }
+    static std::vector<UInt> synToDec, synToInc; // srcCellIdx
+    static std::vector<UInt> inactiveSegmentIndices, activeSegmentIndices; // syn idx
+    removed.clear();                        // purge residual data
```

Comment:
@scottpurdy: This comment is inherited from existing code that I factored out into this function. There is this ugly thing going on in the code, whereby the original variables are declared as `static` in the function (for performance optimization, as @subutai explained it). Since they are static, they retain residual data from the most recent time they are used. This relates to the recent non-thread-safe code discussion.